penrose-web/src/engine/Evaluator.ts

```diff
@@ -252,44 +260,23 @@ export const evalExpr = (
 
     case "AFloat": {
       if (e.contents.tag === "Vary") {
+        console.error("expr", e, "trans", trans, "varyingVars", varyingVars);
         throw new Error("encountered an unsubstituted varying value");
       } else {
         const val = e.contents.contents;
 
+        // Don't convert to VarAD if it's already been converted
         return {
           tag: "Val",
           // Fixed number is stored in translation as number, made differentiable when encountered
           contents: {
             tag: "FloatV",
-            contents: constOf(val),
+            contents: val.tag ? val : constOf(val),
```

Comment:
Can we encode "already converted floats" as a type?

Reply:
I thought that's what this line is doing -- if val is tagged (i.e. is `VarAD`), then don't do anything (because it's an already converted float), otherwise convert it (because it has type `number`).